mixin-metrics.go

```diff
@@ -0,0 +1,319 @@
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".panels[]?.targets[]?.expr"); err != nil {
+		return nil, err
+	}
+
+	if err := ParseJq(&queries, res, ".rows[].panels[].targets[].expr"); err != nil {
+		return nil, err
+	}
+
+	for _, query := range queries {
+		if err := ParseQuery(query, metrics); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+	return &metricsFile, nil
+}
+
+// todo: separate rules and raw metrics
+// parses through a rules file and extracts queries, and then metrics from queries
+func ParseRules(file *os.File) (*MetricsFile, error) {
+
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+
+	var conf RuleConfig
+
+	rulesFile, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(rulesFile, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := conf.RuleGroups
+	for _, group := range groups {
+		for _, rule := range group.Rules {
+			err := ParseQuery(rule.Expr, metrics)
+			if err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	metricsFile := NewMetricsFile(file.Name(), metrics, errors)
+
+	return &metricsFile, nil
+}
+
+// iterates over all rules/dash files in dir
+func ParseDir(dir string, isRules bool) (*MetricsDir, error) {
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	metricsDir := MetricsDir{
+		MetricsFiles: make([]MetricsFile, 0),
+	}
+
+	for _, fileInfo := range files {
+		fmt.Println("Parsing: ", fileInfo.Name())
+		file, err := os.Open(dir + "/" + fileInfo.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		if isRules {
+			metricsFile, err := ParseRules(file)
+			if err != nil {
+				return nil, err
+			}
+			metricsDir.MetricsFiles = append(metricsDir.MetricsFiles, *metricsFile)
+		} else {
+			metricsFile, err := ParseDashboard(file)
+			if err != nil {
+				return nil, err
+			}
+			metricsDir.MetricsFiles = append(metricsDir.MetricsFiles, *metricsFile)
+		}
+
+		err = file.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &metricsDir, nil
+
+}
+
+// write out MetricsDir to outputFile
+func (md *MetricsDir) WriteOut(outputFile string) error {
+	out, err := json.MarshalIndent(*md, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(outputFile, out, os.FileMode(int(0666))); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// prints parsed metrics in relabel_configs form
+func (md *MetricsDir) PrintMetrics() {
+
+	metrics := map[string]struct{}{}
+
+	for _, metricsFile := range md.MetricsFiles {
+		for _, metric := range metricsFile.Metrics{
+			metrics[metric] = struct{}{}
+		}
+	}
+
+	keys := make([]string, 0, len(metrics))
+	for k := range metrics {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	fmt.Println(strings.Join(keys, " | "))
+}
+
+// todo: running `rules` on a dash dir doesn't fail? :(
+func main() {
```

Comment:
nit: if you re-order functions in call order, it's easier to read the code from top to bottom and follow along with what's going on